src/System.Buffers.Experimental/System/Buffers/MemoryList.cs

```diff
@@ -0,0 +1,115 @@
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public class MemoryList : IMemoryList<byte>
+    {
+        private Memory<byte> _data;
+        private MemoryList _next;
+        private long _virtualIndex;
+
+        public MemoryList(Memory<byte> bytes)
+        {
+            _data = bytes;
+            _next = null;
+            _virtualIndex = 0;
+        }
+
+        private MemoryList(Memory<byte> bytes, long runningIndex)
+        {
+            _data = bytes;
+            _virtualIndex = runningIndex;
+        }
+
+        public MemoryList Append(Memory<byte> bytes)
+        {
+            if (_next != null) throw new InvalidOperationException("Node cannot be appended");
+            var node = new MemoryList(bytes, _virtualIndex + _data.Length);
+            _next = node;
+            return node;
+        }
+
+        public Memory<byte> Memory => _data;
+
+        public IMemoryList<byte> Rest => _next;
+
+        public long VirtualIndex => _virtualIndex;
+
+        public int CopyTo(Span<byte> buffer)
+        {
+            int copied = 0;
+            Position position = default;
+            var free = buffer;
+            while (TryGet(ref position, out ReadOnlyMemory<byte> segment, true))
+            {
+                if (segment.Length > free.Length)
+                {
+                    segment.Span.Slice(0, free.Length).CopyTo(free);
+                    copied += free.Length;
+                }
+                else
+                {
+                    segment.Span.CopyTo(free);
+                    copied += segment.Length;
+                }
+                free = buffer.Slice(copied);
+                if (free.Length == 0) break;
+            }
+            return copied;
+        }
+
+        public bool TryGet(ref Position position, out ReadOnlyMemory<byte> item, bool advance = true)
+        {
+            var result = TryGet(ref position, out Memory<byte> memory, advance);
+            item = memory;
+            return result;
+        }
+
+        public bool TryGet(ref Position position, out Memory<byte> item, bool advance = true)
+        {
+            if (position == default)
+            {
+                item = _data;
+                if (advance)
+                {
+                    position.SetItem(_next);
+                }
+                return (!_data.IsEmpty || _next != null);
+            }
+            else if (position.IsEnd)
+            {
+                item = default;
+                return false;
+            }
+
+            var sequence = position.GetItem<MemoryList>();
+            item = sequence._data;
+            if (advance) { position.SetItem(sequence._next); }
+            return true;
+        }
+
+        public static (MemoryList list, long length) Create(params byte[][] buffers)
+        {
+            if(buffers.Length == 0 || (buffers.Length == 1 && buffers[0].Length == 0))
```

Comment:
nit: spacing

Reply:
Can you explain this check? `(buffers.Length == 1 && buffers[0].Length == 0)`

What happens if buffers.Length > 1 but all buffers[i].Length == 0 for all i?

Reply:
this seems like the right spacing, i.e. it's the default VS editor formatting

Reply:
`if(buffers.Length == 0` -> `if (buffers.Length == 0`